pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -626,6 +626,11 @@ func (s *DiscoveryServer) initConnectionNode(discReq *xdsapi.DiscoveryRequest, c
 	if err := nt.SetServiceInstances(s.Env); err != nil {
 		return err
 	}
+	// If the proxy has no service instances and its a gateway, kill the XDS connection as we cannot
+	// serve any gateway config if we dont know the proxy's service instances
+	if nt.Type == model.Router && (nt.ServiceInstances == nil || len(nt.ServiceInstances) == 0) {
```

Comment:
when is `nt.ServiceInstances == nil` ? It does not look like it possible.

perhaps wait a little bit and retry `SetServiceInstances()` before giving up?

Reply:
There is no definite  duration before giveup.

Reply:
is this something we want to enable for the proxy sidecar too in addition to the router type?

Reply:
No.. proxy sidecar can function without service instances

Reply:
/lgtm

Reply:
This solves the issue that we have seen most often, which is with gateway.
Sidecars *with services* can also face the same issue of pilot not sending any inbound listeners. I have not seen it with sidecars though, is that because a full push to sidecars happens more often ?  

Waiting and retrying may not be a bad idea. It will not solve the problem, but it will reduce the likelihood.